File: android/app/src/main/java/com/noop/analytics/LevelBaselines.kt
package com.noop.analytics

import kotlin.math.sqrt

// MARK: - The baselines, and why they are frozen
//
// Each metric's centre, spread and range, derived ONCE from whatever history the wearer had when the
// level first ran, and then never touched again.
//
// THIS IS THE WHOLE POINT. A rolling baseline makes you run to stand still: improve, and your own mean
// rises with you, your z-score falls back toward zero, and the level hands the gain straight back. Two
// people with identical bodies would also sit at different levels because they had different months
// behind them, and the same wearer's level 70 would mean one thing in March and another in December.
// Frozen, a level is a fixed yardstick — the number moves only when the body does.
//
// THE COST, STATED PLAINLY. A frozen scale drifts out of date. Someone who trains for two years will
// press against the top of a range set when they were unfit, and the level stops discriminating at the
// high end. That is the trade the wearer asked for, and it is the right one for a score meant to be
// compared with itself over time — but it is a real cost, not a free lunch, and [needsRefresh] exists
// so a deliberate, explicit re-derivation is possible without the scale silently sliding.
//
// OUTLIERS MATTER MORE HERE, NOT LESS. Because the range is permanent, one corrupt reading at freeze
// time would define the scale forever. The range therefore comes from percentiles rather than the
// literal smallest and largest samples — see [RANGE_LOW_PCT].

/**
 * One metric's frozen reference: where its middle is, how far it usually swings, and the ends of its
 * useful range.
 *
 * [mean] and [sd] drive the z-score path used by HRV, RHR, VO2max and respiratory rate. [min] and
 * [max] drive the range path used by muscle load, whose raw unit (kilograms of volume load) has no
 * meaningful centre to deviate from — a wearer's zero is a rest day, not a bad reading.
 */
data class Baseline(
    val mean: Double,
    val sd: Double,
    val min: Double,
    val max: Double,
) {
    /** Guards a degenerate spread: a wearer whose readings never move would otherwise divide by zero. */
    val safeSd: Double get() = if (sd > 1e-6) sd else 1.0

    /** Guards a degenerate range, for the same reason. */
    val safeSpan: Double get() = (max - min).let { if (it > 1e-6) it else 1.0 }

    /** Where [value] sits in the frozen range, 0–100. Clipped: the range is fixed, readings are not. */
    fun position(value: Double): Double =
        (((value - min) / safeSpan) * 100.0).coerceIn(0.0, 100.0)
}

/** Which metrics the level reads. Named so a missing one can be reported by name. */
enum class LevelMetric {
    SLEEP_SCORE,
    SLEEP_CONSISTENCY,
    HRV,
    RHR,
    VO2MAX,
    RESP_RATE,
    MUSCLE_LOAD,
    STRESS,
}

object LevelBaselines {

    /**
     * How many readings a metric needs before its own history is used instead of the table.
     *
     * Two weeks. Below that a mean is dominated by whichever fortnight happened to be sampled, and a
     * spread estimated from a handful of nights makes every reading look extreme — and since the
     * result is FROZEN, a bad estimate here is permanent.
     */
    const val MIN_SAMPLES = 14

    /**
     * The percentiles that define a metric's range.
     *
     * Not the literal min and max: the range is frozen forever, so one corrupt reading on the day it is
     * derived would set the scale for good. The 5th and 95th percentiles cut the tails while still
     * spanning what the wearer actually does.
     */
    const val RANGE_LOW_PCT = 5.0
    const val RANGE_HIGH_PCT = 95.0

    /**
     * The fallback table, from the specification. Used ONLY for a metric with too little history at
     * freeze time; its range is taken as ±2 SD, which is where the z-score path saturates anyway.
     */
    val DEFAULT: Map<LevelMetric, Baseline> = mapOf(
        LevelMetric.SLEEP_SCORE to table(75.0, 12.0),
        LevelMetric.SLEEP_CONSISTENCY to table(70.0, 15.0),
        LevelMetric.HRV to table(50.0, 15.0),
        LevelMetric.RHR to table(60.0, 10.0),
        LevelMetric.VO2MAX to table(45.0, 8.0),
        LevelMetric.RESP_RATE to table(16.0, 3.0),
        LevelMetric.MUSCLE_LOAD to table(50.0, 20.0),
        LevelMetric.STRESS to table(40.0, 15.0),
    )

    private fun table(mean: Double, sd: Double) =
        Baseline(mean = mean, sd = sd, min = mean - 2 * sd, max = mean + 2 * sd)

    /**
     * Derive one metric's baseline from the readings available at freeze time.
     *
     * The SD is the population form (divide by n), matching the spec's numpy default — the sample form
     * would put the two platforms a fraction apart on every score, which is exactly the drift the
     * parity contract exists to catch.
     */
    fun derive(metric: LevelMetric, history: List<Double>): Baseline {
        val xs = history.filter { it.isFinite() }.sorted()
        if (xs.size < MIN_SAMPLES) return DEFAULT.getValue(metric)
        val mean = xs.sum() / xs.size
        val variance = xs.sumOf { (it - mean) * (it - mean) } / xs.size
        return Baseline(
            mean = mean,
            sd = sqrt(variance),
            min = percentile(xs, RANGE_LOW_PCT),
            max = percentile(xs, RANGE_HIGH_PCT),
        )
    }

    /**
     * Derive every metric at once. What a caller does exactly once, and then stores.
     *
     * A metric absent from [history] keeps the table entry rather than being dropped: the set of
     * baselines has to be complete, or a component would silently stop scoring the day a sensor came
     * online.
     */
    fun deriveAll(history: Map<LevelMetric, List<Double>>): Map<LevelMetric, Baseline> =
        LevelMetric.entries.associateWith { derive(it, history[it].orEmpty()) }

    /**
     * Linear-interpolated percentile over an ALREADY SORTED list.
     *
     * Interpolated rather than nearest-rank so the two platforms cannot land on different samples at
     * the same percentile — with the result frozen, a one-sample disagreement would be permanent.
     */
    internal fun percentile(sorted: List<Double>, pct: Double): Double {
        if (sorted.isEmpty()) return 0.0
        if (sorted.size == 1) return sorted[0]
        val rank = (pct / 100.0) * (sorted.size - 1)
        val lo = rank.toInt()
        val hi = (lo + 1).coerceAtMost(sorted.size - 1)
        val frac = rank - lo
        return sorted[lo] + (sorted[hi] - sorted[lo]) * frac
    }

    /**
     * Whether a frozen set is thin enough to be worth re-deriving.
     *
     * NOT automatic. A caller may offer the wearer a re-derivation — "your scale was set from two weeks
     * of data, reset it against the year you now have" — but nothing here does it on its own, because
     * a scale that moves by itself is the thing this whole file exists to prevent.
     */
    fun needsRefresh(frozen: Map<LevelMetric, Baseline>): Boolean =
        LevelMetric.entries.count { frozen[it] == DEFAULT[it] } >= LevelMetric.entries.size / 2
}

File: Packages/StrandAnalytics/Sources/StrandAnalytics/LevelBaselines.swift
import Foundation

// LevelBaselines.swift — the baselines, and why they are frozen.
//
// iOS lane. This scale no longer matches the Android `LevelBaselines`: the level was rebuilt on iOS to
// have no ceiling (see `LevelEngine`), and the Android lane was deliberately left as it is.
//
// Each metric's centre, spread and useful range, derived from the wearer's own history and then never
// touched again.
//
// THIS IS THE WHOLE POINT. A rolling baseline makes you run to stand still: improve, and your own mean
// rises with you, your score falls back toward the middle, and the level hands the gain straight back.
// Frozen, a level is a fixed yardstick — the number moves only when the body does. And with no ceiling
// on the scale, a body that outgrows its old range keeps scoring above 100 rather than pressing against
// a cap.
//
// THE RANGE IS THE 5th AND 95th PERCENTILE, not the literal smallest and largest samples. The top of the
// range is what a score of 100 MEANS — "your own 95th-percentile day" — and one corrupt reading at
// freeze time would otherwise define 100 for good, somewhere the wearer can never reach again.

/// One metric's frozen reference.
///
/// `mean` is where a score of 50 sits. `max` (the 95th percentile) is 100 for a metric where higher is
/// better; `min` (the 5th) is 100 where lower is better. Nothing is clipped: a reading beyond the range
/// scores beyond 100, and one far below the mean scores below 0.
public struct Baseline: Equatable, Sendable {
    public let mean: Double
    public let sd: Double
    public let min: Double
    public let max: Double

    public init(mean: Double, sd: Double, min: Double, max: Double) {
        self.mean = mean
        self.sd = sd
        self.min = min
        self.max = max
    }

    /// Guards a degenerate spread: a wearer whose readings never move would otherwise divide by zero.
    public var safeSd: Double { sd > 1e-6 ? sd : 1 }

    /// The score of `value`: 50 at the mean, 100 at the wearer's 95th-percentile day in the good
    /// direction, linear either side, and UNBOUNDED both ways.
    ///
    /// When the good end of the range sits on the mean — a metric that barely moved at freeze time —
    /// the distance is taken as 1.645 SD instead, which is where the 95th percentile of a normal spread
    /// would be, so the scale still has a sensible slope.
    public func score(_ value: Double, higherIsBetter: Bool) -> Double {
        let best = higherIsBetter ? max : min
        var span = best - mean
        if abs(span) < 1e-6 { span = (higherIsBetter ? 1 : -1) * 1.645 * safeSd }
        return 50 + 50 * (value - mean) / span
    }
}

/// Which metrics the level reads. Named so a missing one can be reported by name.
public enum LevelMetric: String, CaseIterable, Sendable, Codable {
    /// Deep + REM minutes a night.
    case restorativeMin
    /// How far bedtime and wake time moved against the night before, in minutes (lower is better).
    case sleepRegularityMin
    case hrv
    case rhr
    case vo2max
    case respRate
    case muscleLoad
    /// Mean RMSSD over the day's still, scored waking hours — daytime calm.
    case daytimeRmssd
}

public enum LevelBaselines {

    /// How many readings a metric needs before its own history is used instead of the table.
    ///
    /// Two weeks. Below that a mean is dominated by whichever fortnight happened to be sampled — and
    /// since the result is FROZEN, a bad estimate here would be permanent.
    public static let minSamples = 14

    public static let rangeLowPct: Double = 5
    public static let rangeHighPct: Double = 95

    /// The fallback table, for a metric with too little history YET. Never frozen — see
    /// `LevelBaselineStore`: a metric on the table is re-derived on every load until its own history
    /// is deep enough, and frozen then.
    public static let table: [LevelMetric: Baseline] = [
        .restorativeMin: entry(mean: 170, sd: 40),
        .sleepRegularityMin: entry(mean: 45, sd: 25),
        .hrv: entry(mean: 50, sd: 15),
        .rhr: entry(mean: 60, sd: 10),
        .vo2max: entry(mean: 45, sd: 8),
        .respRate: entry(mean: 16, sd: 3),
        .muscleLoad: entry(mean: 5000, sd: 2500),
        .daytimeRmssd: entry(mean: 35, sd: 12),
    ]

    private static func entry(mean: Double, sd: Double) -> Baseline {
        Baseline(mean: mean, sd: sd, min: mean - 1.645 * sd, max: mean + 1.645 * sd)
    }

    /// Whether `history` is deep enough to freeze this metric.
    public static func isDerivable(_ history: [Double]) -> Bool {
        history.filter(\.isFinite).count >= minSamples
    }

    /// Derive one metric's baseline, or its table entry when the history is too thin.
    ///
    /// The SD is the POPULATION form (divide by n).
    public static func derive(_ metric: LevelMetric, history: [Double]) -> Baseline {
        let xs = history.filter { $0.isFinite }.sorted()
        guard xs.count >= minSamples else { return table[metric] ?? entry(mean: 50, sd: 15) }
        let mean = xs.reduce(0, +) / Double(xs.count)
        let variance = xs.reduce(0) { $0 + ($1 - mean) * ($1 - mean) } / Double(xs.count)
        return Baseline(
            mean: mean,
            sd: variance.squareRoot(),
            min: percentile(xs, rangeLowPct),
            max: percentile(xs, rangeHighPct)
        )
    }

    public static func deriveAll(history: [LevelMetric: [Double]]) -> [LevelMetric: Baseline] {
        var out: [LevelMetric: Baseline] = [:]
        for metric in LevelMetric.allCases {
            out[metric] = derive(metric, history: history[metric] ?? [])
        }
        return out
    }

    /// Linear-interpolated percentile over an ALREADY SORTED array.
    public static func percentile(_ sorted: [Double], _ pct: Double) -> Double {
        guard !sorted.isEmpty else { return 0 }
        guard sorted.count > 1 else { return sorted[0] }
        let rank = (pct / 100) * Double(sorted.count - 1)
        let lo = Int(rank)
        let hi = Swift.min(lo + 1, sorted.count - 1)
        let frac = rank - Double(lo)
        return sorted[lo] + (sorted[hi] - sorted[lo]) * frac
    }
}
